R/assign.R

```diff
@@ -0,0 +1,473 @@
+    # polygons if they are close enough to one another (user-defined)
+    datFirst$genetID <- groupByGenet(datFirst, buffGenet) # put these temporary
+    # genet ID's in a temporary 'genetID' column, which will tell us to give
+    # these genets the same track ID
+    ## aggregate size by genetID (total size for each ramet)
+    datFirst <- st_join(datFirst, (datFirst %>%
+      group_by(genetID) %>%
+      summarize(rametAreaTemp = sum(Area))), by = "genetID") %>%
+      select(-c(genetID.x, rametArea)) %>%
+      rename("genetID" = "genetID.y", "rametArea" = "rametAreaTemp")
+  }
+  if (clonal == 0) { ## if genets are not allowed (clonal == 0), then each
+    # individual polygon gets a unique value in the 'genetID' column
+    datFirst$genetID <- 1:nrow(datFirst)
+    datFirst$rametArea <- datFirst$Area
+  }
+
+  ## assign a unique trackID to every unique genetID in this first-year dataset
+  IDs <- data.frame( "genetID" = sort(unique(datFirst$genetID)), ## get a vector
+                     # of all of the unique genetIDs
+            "trackID" = paste0(unique(dat$sp_code_6), ## get the unique 6-letter
+                               # species code
+                               "_",unique(datFirst$Year), ## get the unique year
+                  "_",c(1:length(unique(datFirst$genetID))))) ## get a vector of
+  # unique numbers that is the same length as the genetIDs in this quad/year
+
+  datFirst<- merge(datFirst[,names(datFirst) != "trackID"], IDs, by = "genetID")
+
+  ## assign the first-year data to the 'tempCurrentYear' data.frame
+  tempCurrentYear <- datFirst ## this data.frame will get redefined for
+  # each iteration of the for-loop below
+
+  ## give all individuals in year #1 a '0' in the ghost column
+  tempCurrentYear$ghost <- 0
+
+  ##  i = year in inventory
+  for (i in 2:length(inv)) {
+    ## CHECK IF YEARS ARE CONTINUOUS -- check to see if the sampling years of
+    # 'tempCurrentYear' and 'tempNextYear' are not far enough apart to exceed
+    # the 'dormancy' argument. If dormancy is not exceeded, then go ahead with
+    # this loop. If it is, then freshly redefine 'tempCurrentYear' and proceed
+    # to the next 'i'
+    if (inv[i] - inv[i-1] <= (dorm+1)) {
+      ## 'tempCurrentYear' is the sf data.frame of the 'current' year
+      ## need to get the sf data.frame of the 'next' year (year 'i')
+      tempNextYear <- dat[dat$Year==inv[i],]
+
+      ## MAKE SURE THERE IS DATA IN YEAR i-1 (tempCurrentYear isn't empty) If
+      # not, then go to the next i (only after replacing 'tempCurrentYear' with
+      # the tempNextYear data.frame)
+      if (nrow(tempCurrentYear)>0) {
+        ## add a buffer to the current year data
+        tempCurrentBuff <- st_buffer(tempCurrentYear, buff)
+
+        ## MAKE SURE THERE IS DATA IN YEAR i (tempNextYear isn't empty) If not,
+        # then go to next i
+        if (nrow(tempNextYear)>0) {
+          ## AGGREGATE BY GENET for year i (if clonal = 1)
+          if(clonal==1) {
```

Comment:
This `if(clonal==1){...}else if(clonal==0){...}` is done twice in this function. You can write a function that lives inside of another function. I call these internal functions. I might be worth it here.